lapis/src/main/kotlin/org/genspectrum/lapis/model/AdvancedQueryCustomListener.kt

```diff
@@ -0,0 +1,516 @@
+            metadataFieldsByName[metadataName.lowercase(Locale.US)]
+                ?: throw BadRequestException("Metadata field $metadataName does not exist", null)
+        when (field.type) {
+            MetadataType.DATE -> {
+                try {
+                    val date = LocalDate.parse(metadataValue)
+                    expressionStack.addLast(DateBetween(field.name, to = null, from = date))
+                } catch (exception: DateTimeParseException) {
+                    throw BadRequestException("'$metadataValue' is not a valid date: ${exception.message}", exception)
+                }
+            }
+
+            MetadataType.FLOAT -> {
+                try {
+                    expressionStack.addLast(FloatBetween(field.name, from = metadataValue.toDouble(), to = null))
+                } catch (e: NumberFormatException) {
+                    throw BadRequestException("'$metadataValue' is not a valid float", e)
+                }
+            }
+
+            MetadataType.INT -> {
+                try {
+                    expressionStack.addLast(IntBetween(field.name, from = metadataValue.toInt(), to = null))
+                } catch (e: NumberFormatException) {
+                    throw BadRequestException("'$metadataValue' is not a valid integer", e)
+                }
+            }
+
+            else -> {
+                throw BadRequestException("expression >= cannot be used for ${field.type}", null)
+            }
+        }
+    }
+
+    private fun mapToLineageFilter(
+        metadataName: String,
+        metadataValue: String?,
+    ): LineageEquals {
+        val sanitizedValue = metadataValue?.removeSurrounding("'")
+
+        return when {
+            sanitizedValue.isNullOrBlank() -> throw BadRequestException(
+                "Invalid lineage: $sanitizedValue is NULL - to search for NULL values use `IsNull($metadataName)`?",
+            )
+            sanitizedValue.endsWith(
+                ".*",
+            ) -> LineageEquals(metadataName, sanitizedValue.substringBeforeLast(".*"), includeSublineages = true)
+            sanitizedValue.endsWith("*") -> LineageEquals(
+                metadataName,
+                sanitizedValue.substringBeforeLast("*"),
+                includeSublineages = true,
+            )
+            sanitizedValue.endsWith('.') -> throw BadRequestException(
+                "Invalid lineage: $sanitizedValue must not end with a dot. Did you mean '$sanitizedValue*'?",
+            )
+            else -> LineageEquals(metadataName, sanitizedValue, includeSublineages = false)
+        }
+    }
+
+    override fun enterMetadataQuery(ctx: AdvancedQueryParser.MetadataQueryContext) {
+        val metadataName = ctx.geneOrName().text
+        val metadataValue = ctx.value().text.trim('\'')
+
+        var name = metadataName
+        if (metadataName.endsWith(".regex")) {
+            name = metadataName.substringBeforeLast(".regex")
+        }
```

Comment:
IMO this yields too convoluted code. This is probably easier to follow:

```suggestion
        if (metadataName.endsWith(".regex")) {
            name = findMetadataField(metadataName.substringBeforeLast(".regex"))

            if (!field.lapisAllowsRegexSearch) {
                throw BadRequestException("Metadata field `${field.name}` does not support regex search.",)
            }

            expressionStack.addLast(StringSearch(field.name, metadataValue))
            return
        }
        val name = metadataName
```

It splits the method into separate dataflows that finish independently (in contrast to having parallel flows, i.e. the regex case and the non-regex case at the same time). Then we can remove the regex cases from the `when`  below.